simpervisor/atexitasync.py

```diff
@@ -8,27 +8,69 @@
 
 _handlers = []
 
+_prev_handlers = {}
 signal_handler_set = False
 
 
 def add_handler(handler):
-    global signal_handler_set
-    if not signal_handler_set:
-        signal.signal(signal.SIGINT, _handle_signal)
-        signal.signal(signal.SIGTERM, _handle_signal)
-        signal_handler_set = True
+    """
+    Adds a signal handler function that will be called when the Python process
+    receives either a SIGINT (on windows CTRL_C_EVENT) or SIGTERM signal.
+    """
+    _ensure_signal_handlers_set()
     _handlers.append(handler)
 
 
 def remove_handler(handler):
+    """Removes previously added signal handler."""
     _handlers.remove(handler)
 
 
+def _ensure_signal_handlers_set():
+    """
+    Ensures _handle_signal is registered as a top level signal handler for
+    SIGINT and SIGTERM and saves previously registered non-default Python
+    callable signal handlers.
+    """
+    global signal_handler_set
+    if not signal_handler_set:
+        # save previously registered non-default Python callable signal handlers
+        #
+        # windows note: signal.getsignal(signal.CTRL_C_EVENT) would error with
+        # "ValueError: signal number out of range", and
+        # signal.signal(signal.CTRL_C_EVENT, _handle_signal) would error with
+        # "ValueError: invalid signal value".
+        #
+        prev_sigint = signal.getsignal(signal.SIGINT)
+        prev_sigterm = signal.getsignal(signal.SIGTERM)
+        if callable(prev_sigint) and prev_sigint.__qualname__ != "default_int_handler":
+            _prev_handlers[signal.SIGINT] = prev_sigint
+        if callable(prev_sigterm) and prev_sigterm != signal.Handlers.SIG_DFL:
+            _prev_handlers[signal.SIGTERM] = prev_sigint
```

Comment:
I think there is a typo here. It should be `prev_sigterm` and not `prev_sigint`

Reply:
Ooooh nice critical catch!